src/librustc/traits/select.rs

```diff
@@ -2630,7 +2630,15 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
                                            snapshot)
         });
 
-        obligations.extend(trait_obligations);
```

Comment:
?

Reply:
does this serve any purpose? If not, please remove it.

Reply:
(note that "measure-free perf hack" is not a valid purpose)

Reply:
It was just a way to avoid possibly multiple allocations-and-copy of the vector.  We know that it is at most `trait_obligations.len()` elements and the vast majority will be `trait_obligations.len() - 1`.

I will remove it, but I personally think it is good practice to pre-allocate vectors if the maximum number of elements is known.

Reply:
pre-allocating is OK. the problem is that you added the `if` check. Please remove it.